Add taproot data structures

sanket1729 opened the pull request:
I am still working on adding tests and testing these changes against bitcoin core. Feedback on code and API design is welcome. 

Supercedes #666 

- Builder API to construct trees
- Huffman encoding API that outputs optimal taproot tree given odds.

TODO:
- [x] Add control block APIs. Verify merkle proofs, create a control block from spend data
- [x] Add tests and verify them against bitcoin consensus

GeneFerneau commented on src/util/sighash.rs:
```diff
@@ -31,6 +31,8 @@ use {Script, Transaction, TxOut};
 
 use prelude::*;
 
+use super::taproot::{LeafVersion};
```
nit:

```suggestion
use super::taproot::LeafVersion;
```

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            let mut start = 0;
+            while start < sl.len() {
+                let elem = sha256::Hash::from_slice(&sl[start..start + TAPROOT_CONTROL_NODE_SIZE])
+                    .expect("must be multiple for 32");
```
This condition is checked at the start of the function. `sl.len() % TAPROOT_CONTROL_NODE_SIZE != 0`, so this panic can never be triggered. 

I would also need to create a new error type for propagating errors that can never occur and could be confusing to the reader.

sanket1729 replied:
> and other places

If you find any reachable panic in this PR, that is a bug.

GeneFerneau replied:
> I would also need to create a new error type for propagating errors that can never occur and could be confusing to the reader.

That makes sense. The large number of `expect` calls just looked odd on first review. Will keep this in mind on further review.

tcharding replied:
Perhaps instead of "must be multiple for 32" we could use "unreachable: we check this above" or something like that. Another option is to have a comment `// Unreachable: checked above.`  - I'd definitely prefer to see some explicit thing so ever dev that reads theses `expect` calls does not have to waste clock cycles wondering why we panic.

GeneFerneau commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+/// Tapscript version
+// https://github.com/bitcoin/bitcoin/blob/e826b22da252e0599c61d21c98ff89f366b3120f/src/script/interpreter.h#L226
+pub const TAPROOT_LEAF_TAPSCRIPT: u8 = 0xc0;
+/// Tapscript version
```
```suggestion
/// Tapscript control base size
```

GeneFerneau commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+    // Internal function to append elements to proof
+    fn push(&mut self, h: sha256::Hash) -> Result<(), TaprootBuilderError> {
+        self.0.push(h);
+        if self.0.len() > TAPROOT_CONTROL_MAX_NODE_COUNT {
+            return Err(TaprootBuilderError::InvalidMerkleTreeDepth(self.0.len()));
+        } else {
+            Ok(())
+        }
```
nit: check for error condition before mutating internal state

```suggestion
        if self.0.len() + 1 > TAPROOT_CONTROL_MAX_NODE_COUNT {
            Err(TaprootBuilderError::InvalidMerkleTreeDepth(self.0.len() + 1))
        } else {
            self.0.push(h);
            Ok(())
        }
```

sanket1729 replied:
Good catch!

Kixunil replied:
Rewriting it to `self.0.len() >= TAPROOT_CONTROL_MAX_NODE_COUNT` would avoid overflow if the invariant that `self.0.len() <= TAPROOT_CONTROL_MAX_NODE_COUNT` is ever accidentally broken.

tcharding reviewed (commented):
A couple more minor comments, thanks.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+                sl.len() / TAPROOT_CONTROL_NODE_SIZE,
+            ))
+        } else {
+            let mut inner = Vec::with_capacity(sl.len() / TAPROOT_CONTROL_NODE_SIZE);
```
Looks like this whole branch could be just:

```rust
let inner = sl
    .chunks_exact(TAPROOT_CONTROL_NODE_SIZE)
    .map(|chunk| sha256::Hash::from_slice(chunk).expect("chunk exact always returns the correct size"))
    .collect();
Ok(TaprootMerkleBranch(inner))
```

One day when MSRV allows for `array_chunks` we could even get rid of unwrap.

sanket1729 replied:
Will use using iterators and chunks for all from_slice methods

Kixunil commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        {
+            return Err(TaprootError::InvalidControlBlockSize(sl.len()));
+        }
+        let output_key_parity = (sl[0] & 1) == 1;
```
Turning the slice into iterator would make it easier to reason about this code IMO, but we really need the extension trait I mentioned somewhere already.

sanket1729 replied:
How do you suggest approaching this? I don't see a way to process an iterator u8 and ask it to yield slices without explicating allocating expected size arrays.

sanket1729 commented:
Sorry for the force push, addressed the feedback from @GeneFerneau and @tcharding, did not address comments from @Kixunil. I will add fixup commits from now on for easier review and later squash them when it is ready.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+                // Can't propagate further up than the root
+                return Err(TaprootBuilderError::OverCompleteTree);
+            }
+            let child = self.branch.pop().unwrap().unwrap(); // checked for size and is_some
```
Even though this is more rustic, I think this is more difficult to review. Maybe I can have one unwrap/expect for the pop and only match on the option part. 

```rust
while self.branch.len() == depth + 1 {
    if depth == 0 {
        // We are trying to combine two nodes at root level.
        // Can't propagate further up than the root
        return Err(TaprootBuilderError::OverCompleteTree);
    }
    let child = match self.branch.pop() {
        None => unreachable!("Len of branch checked to be >= 1"),
        Some(Some(child)) => child,
        Some(None) => break, // Cannot combine further as node is None
    };
    node = NodeInfo::combine(node, child)?;
    // Propagate to combine nodes at a lower depth
    depth -= 1;
}
```

sanket1729 replied:
I ended up implementing this in the next commit.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -558,14 +558,13 @@ impl TaprootMerkleBranch {
-                start += TAPROOT_CONTROL_NODE_SIZE;
-            }
+            let inner = sl
+                .chunks(TAPROOT_CONTROL_NODE_SIZE)
```
`chunks_exact` is faster though since we know the size...

sanket1729 replied:
But it is rust 1.31 :(, outside our current MSRV of 1.29

sanket1729 replied:
Added to the list in #510

RCasatta commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,772 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        let mut bytes_written = 0;
+        bytes_written += writer.write(&[first_byte])?;
+        bytes_written += writer.write(&self.internal_key.serialize())?;
+        bytes_written += writer.write(&self.merkle_branch.serialize())?;
```
```suggestion
        bytes_written += self.merkle_branch.serialize_to_writer(&mut writer)?;
```

GeneFerneau commented on src/util/taproot.rs:
```diff
@@ -169,4 +920,105 @@ mod test {
+            let ctrl_block = tree_info.control_block(&ver_script).unwrap();
+            assert!(ctrl_block.verify_taproot_commitment(&secp, &output_key, &ver_script.0))
+        }
+    }
```
I mean in the case where there are no "real" scripts, but only the unspendable script suggested by [BIP 341](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#constructing-and-spending-taproot-outputs):

> If the spending conditions do not require a script path, the output key should commit to an unspendable script path instead of having no script path. This can be achieved by computing the output key point as Q = P + int(hashTapTweak(bytes(P)))G

Not sure how opinionated this library should be about that recommendation.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,772 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        let mut output_key = internal_key;
+        // # Panics:
+        //
+        // This would return Err if the merkle root hash is the negation of the secret
+        // key corresponding to the internal key.
+        // As this is unlikely to occur (1/2^128) in real life usage, it is safe to unwrap this
+        // if the merkle root is derived as specified in BIP341
+        let parity = output_key
+            .tweak_add_assign(&secp, &tweak)
+            .expect("Cryptographically unreachable code");
```
This is because the tweak is either `H(P)` or `H(P||S)` both of which are random outputs and hence it crafting such a merkle should be impossible.

sanket1729 replied:
I think the comment here is incorrect as it mixes the usage of tweak and merkle root. Add some clarification in fccd4a8659fa68b251fec47e28078f84f5b1a36c

apoelstra commented on src/util/taproot.rs:
```diff
@@ -82,6 +90,243 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+pub struct ControlBlock {
+    /// The internal key
+    pub internal_key: schnorr::PublicKey,
+    /// The parity of the output key(NOT THE INTERNAL KEY WHICH IS ALWAYS XONLY)
```
nit: space before `(`. I really appreciate this comment though.

apoelstra commented on src/util/taproot.rs:
```diff
@@ -82,6 +90,243 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+}
+
+impl ControlBlock {
+    /// Obtain a ControlBlock from size. This is an extra witness element
```
nit: s/size/a slice/

GeneFerneau commented:
tACK [38a601d](https://github.com/rust-bitcoin/rust-bitcoin/pull/677/commits/38a601d59138007bfb227c7f2361161aeeecde32)

Will start writing some code against this PR's API, and respond with feedback.

apoelstra commented on src/util/taproot.rs:
```diff
@@ -146,6 +146,49 @@ pub struct TaprootSpendInfo {
+        // Therefore, the loop will eventually terminate with exactly 1 element
+        assert!(node_odds.len() == 1);
+        let node = node_odds.pop().unwrap().1;
+        return Ok(Self::from_node_info(secp, internal_key, node));
```
I see, I was mistaken about the ability to move out of arrays with indexing.

sanket1729 commented:
Rebased and squashed. No diff between 38a601d and cb00856

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -110,6 +111,108 @@ pub const TAPROOT_CONTROL_BASE_SIZE: usize = 33;
+    output_key_parity: bool,
+    /// The tweaked output key
+    output_key: schnorr::PublicKey,
+    /// Map from (script, leaf_version) to (sets of) control blocks.
```
not to a set of control blocks but to a merkle proof of the script

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+/// Control Block data structure used in Tapscript satisfaction
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ControlBlock {
+    /// The internal key
+    pub internal_key: schnorr::PublicKey,
+    /// The parity of the output key (NOT THE INTERNAL KEY WHICH IS ALWAYS XONLY)
+    pub output_key_parity: bool,
+    /// The tapleaf version,
+    pub leaf_version: LeafVersion,
+    /// The merkle proof of a script associated with this leaf
+    pub merkle_branch: TaprootMerkleBranch,
```
I would prefer to have the same field ordering as in serialization: leaf version, parity, internal key and merkle path

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+/// The leaf version for tapleafs
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LeafVersion {
+    ver: u8,
```
nit: why do we need a named field here, not just `(u8)`?

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+impl LeafVersion {
+    /// Obtain LeafVersion from u8, will error when the last bit of ver is not 0
+    pub fn from_u8(ver: u8) -> Result<Self, TaprootError> {
+        if ver & TAPROOT_LEAF_MASK != ver {
+            Err(TaprootError::InvalidTaprootLeafVersion(ver))
+        } else {
+            Ok(LeafVersion { ver: ver })
```
This allows construction of an incorrect leaf versions (like 0x50 etc) which will conflict with annex etc. I think we need enumeration of all allowed leaf versions and check against it.

dr-orlovsky replied:
here the list of the values from the BIP-341:

> The values that comply to this rule are the 32 even values between 0xc0 and 0xfe and also 0x66, 0x7e, 0x80, 0x84, 0x96, 0x98, 0xba, 0xbc, 0xbe

– <https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_note-7>

sanket1729 replied:
Good catch about the values. I knew about this rule, but somehow forgot to implement it :(

sanket1729 replied:
Actually, we only need to test for the annex tag. Other values are only recommendations and not consensus rules.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// Invalid Control Block Size
+    InvalidControlBlockSize(usize),
+    /// Invalid taproot internal key
+    InvalidInternalKey(secp256k1::Error),
```
We need either `source` on this type - or just skip the inner error since it will always be `secp256k1::Error::InvalidSecretKey` without adding any useful information – but making enumeration of error types by library users more complex.

sanket1729 replied:
The idea is to add `source` when we have it in MSRV. Since the MSRV change is expected soon, I think it's okay to have it is right now. I have added a TODO comment stating the same.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+/// Detailed error type for taproot utilities
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum TaprootError {
```
Needs `impl std::error::Error` for `std` feature

sanket1729 replied:
It is implemented below

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+/// Detailed error type for taproot builder
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum TaprootBuilderError {
```
Needs `impl std::error::Error` for `std` feature

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,776 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// Two nodes at depth 0 are not allowed
+    OverCompleteTree,
+    /// Invalid taproot internal key
+    InvalidInternalKey(secp256k1::Error),
```
See my comment on `TaprootError` for similar variant there

sanket1729 replied:
Answered this in another comment there. I think we are more likely to add the source if we keep the error type

dr-orlovsky reviewed (changes requested):
Did an initial review on the data type structure against the spec; will process with business logic review a bit later. Left few notes

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +92,753 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            }
+            Ok(TaprootMerkleBranch(inner))
+        }
+    }
```
I believe the current way of writing this is better. Willing to change it to this if you think there is a serious benefit here.

tcharding replied:
No serious benefit, just ideas. You judgement is ok with me :)

sanket1729 commented:
Addressed all your comments @dr-orlovsky

dr-orlovsky commented:
Will complete full review today

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        script_weights: I,
+    ) -> Result<Self, TaprootBuilderError>
+    where
+        I: IntoIterator<Item = (usize, Script)>,
```
I am still tempting to say that we should have a distinct `TapScript` type, such that we can be sure that the Script data are valid by type safety... Nevermind, this is probably not a thing for this PR

Kixunil replied:
I agree their semantics is sufficiently different to use separate types.

sanket1729 replied:
I agree that separating scripts based on context (script pubkey vs script sig vs witness script vs tapscript) definitely has value. We can make sure address APIs are called correctly on each script type. That is a larger undertaking that would require refactoring the entire codebase. Best left for later.

Kixunil replied:
Sure, agree with doing it later.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        I: IntoIterator<Item = (usize, Script)>,
+        C: secp256k1::Verification,
+    {
+        let mut node_weights = BinaryHeap::<(usize, NodeInfo)>::new();
```
If we do allocation anyway, why do we need an iterator and not taking as an input, `BTreeMap<Script, usize>` ensuring that each script spending path is unique?

sanket1729 replied:
Each spending script does not have to be unique. Taproot allows repeated scripts. Which is why the current API is IntoInterator<Script, size> that also allows Vec<(script, usize)>

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        for (p, leaf) in script_weights {
+            node_weights.push((p, NodeInfo::new_leaf_with_ver(leaf, LeafVersion::default())));
+        }
+        if node_weights.is_empty() {
```
we can have this check at the start of the function, avoiding unnecessary allocs

sanket1729 replied:
Since this is an iterator, I cannot check with allocations. But, does Vec::new() heap allocate anything?  All the solutions that I can think of (making it peekable) or matching the first element, look ugly to me. and not worth the tradeoff here.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        }
+        while node_weights.len() > 1 {
+            // Combine the last two elements and insert a new node
+            let (p1, s1) = node_weights.pop().expect("len must be atleast two");
```
```suggestion
            let (p1, s1) = node_weights.pop().expect("len must be at least one");
```

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        while node_weights.len() > 1 {
+            // Combine the last two elements and insert a new node
+            let (p1, s1) = node_weights.pop().expect("len must be atleast two");
+            let (p2, s2) = node_weights.pop().expect("len must be atleast two");
```
```suggestion
            let (p2, s2) = node_weights.pop().expect("len must be at least two");
```

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            let (p1, s1) = node_weights.pop().expect("len must be atleast two");
+            let (p2, s2) = node_weights.pop().expect("len must be atleast two");
+            // Insert the sum of first two in the tree as a new node
+            node_weights.push((p1 + p2, NodeInfo::combine(s1, s2)?));
```
Alternative without `expect`:

``` rust
if let (Some(n1), Some(n2)) = (node_weights.pop(), node_weights.pop()) {
   node_weights.push(n1.0 + n2.0, NodeInfo::combine(n1.1, n2.1)?);
}
```

sanket1729 replied:
I think this the same as before, with unreachable! in else part.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        }
+        // Every iteration of the loop reduces the node_weights.len() by exactly 1
+        // Therefore, the loop will eventually terminate with exactly 1 element
+        assert!(node_weights.len() == 1);
```
rust stdlib prefers to use in such cases `debug_assert!`

sanket1729 replied:
What is `such cases` here? 
From https://doc.rust-lang.org/std/macro.debug_assert.html: 

Replacing assert! with debug_assert! is thus only encouraged after thorough profiling, and more importantly, only in safe code!

dr-orlovsky replied:
debug assert is used when some statement may be broken only because of the language bug (for instance, 1-element array does not pop) - this helps to optimize production code (such asserts are removed from the release builds)

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        // Every iteration of the loop reduces the node_weights.len() by exactly 1
+        // Therefore, the loop will eventually terminate with exactly 1 element
+        assert!(node_weights.len() == 1);
+        let node = node_weights.pop().unwrap().1;
```
two previous line can be just one with `let node = node_weights.pop().expect("huffman tree algorithm is broken").1;`

Kixunil replied:
+1 on using `except()` instead of `unwrap()`

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        return Ok(Self::from_node_info(secp, internal_key, node));
+    }
+
+    /// Create a new key spend with internal key and proided merkle root.
```
```suggestion
    /// Create a new key spend with internal key and provided merkle root.
```

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    /// Create a new key spend with internal key and proided merkle root.
+    /// Provide [`None`] for merkle_root if there is no script path.
```
```suggestion
    /// Provide [`None`] for merkle_root if there are no script path spending conditions.
```

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+    /// Create a new key spend with internal key and proided merkle root.
+    /// Provide [`None`] for merkle_root if there is no script path.
+    /// # *Note*: As per BIP341
```
If this should be a title, we rustdoc a line break before. But probably just the first `#` is unnecessary

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// Refer to BIP 341 footnote (Why should the output key always have
+    /// a taproot commitment, even if there is no script path?) for more details
+    ///
+    pub fn new_key_spend<C: secp256k1::Verification>(
```
Rust API recommends to name such functions `with_key_spend`

sanket1729 replied:
I am still not super comfortable with `with_stuff` vs `new_stuff`. Plus, we already all APIs in rust-bitcoin are with `new_`, 

In this particular case, I think new_key_spend makes more sense vs `with_key_spend`. I think `with` makes more sense for simpler operations that directly operate on a field like `Vec::with_capacity()`. Maybe `with_internal_key`? I somewhat prefer `new_key_spend`, but I will go with what most reviewers/users want.

dr-orlovsky replied:
`with_internal_key` seems more reasonable, but I will not ask to change just a function name since this is nit

apoelstra replied:
I don't have a strong opinion on this, but FWIW I think the correct thing is `with_` and I might consider fixing all the other instances of `new_` in a future mega-breaking API modernization PR.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    ) -> Self {
+        let tweak = compute_tap_tweak(merkle_root, internal_key);
+        let mut output_key = internal_key;
+        // # Panics:
```
This probably needs to be a part of the docs, i.e. moved to the header + replaced with three slashes

Kixunil replied:
This looks like internal panic, which typically aren't documented but TBH I didn't have time to look deeply.

sanket1729 replied:
I just wanted to highlight this with an explanation to get a careful review about that particular `expect`.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        // this is unlikely to occur (1/2^128) in real life usage, it is safe to unwrap this
+        let parity = output_key
+            .tweak_add_assign(&secp, &tweak)
+            .expect("Cryptographically unreachable code");
```
```suggestion
            .expect("compute_tap_tweak functions is broken");
```

This has much higher probability than SHA-256 collision or a tweak going above curve order :D

sanket1729 replied:
Last I remember, I checked the libsecp256k1 API to make sure that `tweak_add_assign` does not when the output is out of order. We can discuss in length the exact security, but I think this is equivalent to 128-bit security. 

The point here is that this is unreachable. I can change it to "BIP 341 TapTweak is broken" if that is more descriptive. I feel the comment above does a good job of explaining why

Kixunil replied:
I think you misunderstood his point which was "since 1/2^128 is impossible, if it actually panicks it means the implementation is broken, so better use a message saying it's broken".

apoelstra replied:
I don't understand the argument. 2^{128} is cryptographically unreachable.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    /// Access the internal script map
+    pub fn as_script_map(&self) -> &BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>> {
```
`BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>>` looks like a good candidate for at least type alias – or newtype with its own API, taking part of the code redundancy load from `TaprootSpendInfo`

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+pub fn compute_tap_tweak(
+    merkle_root: Option<sha256::Hash>,
+    internal_key: schnorr::PublicKey,
+) -> sha256::Hash {
```
Why we do not return `TapTweakHash` here?

sanket1729 replied:
I should, I will change the methods to operate on correct hashes, taptweakhash, tapbranchhash etc.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        internal_key: schnorr::PublicKey,
+        merkle_root: Option<sha256::Hash>,
+    ) -> Self {
+        let tweak = compute_tap_tweak(merkle_root, internal_key);
```
This really looks like a function on `TapTweakHash` rather than something global

sanket1729 replied:
Yes, will add it there.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        let smallest = merkle_branch_set
+            .iter()
+            .min_by(|x, y| x.0.len().cmp(&y.0.len()))
+            .expect("Non-empty iterator");
```
This will panic for non-script structures. We may assume we guarantee that the script is always present, but if I will set `x.merkle_root = None` (a public writable field) than there is no script, but the index is still there, and while we do not panic, we will produce incorrect control block. So either all fields must be private (modifiable via functions) - nor we should not assume that we always have a script and get rid of indexes

sanket1729 replied:
Indeed, changing Merkle root and internal key to non-pub should fix all of this.

sanket1729 replied:
In which case, I think the expect here would be okay? 

I like having assertions/expect in code because they help reason about algorithmic invariants. If the map has a key the value must have at least one element.

Kixunil replied:
Yes, if invariant is guaranteed by the API, `except()` is OK, although I'd like the message to be something like "broken invariant that merkle_branch_set is non-empty".

dr-orlovsky reviewed (changes requested):
Sorry for a plenty of comments, just like to ensure that I have full understanding of how taproot is going to be implemented in rust-bitcoin. And so far I have a lot of questions. Sorry for taking your time.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+/// Taproot can be spent be either:
+/// - Spending using the key path i.e., with secret key corresponding to the output_key
+/// - By satisfying any of the scripts in the script spent path. Each script can be satisfied by providing
+///   a witness stack consisting of the script's inputs, plus the script itself and the control block.
```
This is correct, but is it worth mentioning it in the doc comment?  Annex is non-standard and does not have any consensus rules associated with it yet. I think mentioning it would only confuse people.

Kixunil replied:
"Note: This library currently does **not** support [annex](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#cite_ref-5-0)"

Should be clear and useful.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// The sign final output pubkey as per BIP 341
+    output_key_parity: bool,
+    /// The tweaked output key
+    output_key: schnorr::PublicKey,
```
This is a bug! Those fields should not be exposed as pubs. I will add accessor methods to it.

apoelstra replied:
These are private fields.

sanket1729 replied:
Earlier the internal key was exposed as a pub, which allowed mutating the internal key without the output key being updated. That was fixed later

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+pub struct TaprootSpendInfo {
+    /// The BIP341 internal key.
+    pub internal_key: schnorr::PublicKey,
+    /// The Merkle root of the script tree (None if there are no scripts)
```
The internal key is tweaked correctly into a separate field output key when there are no scripts. Users should use output-key as the tweaked key. This handles both cases, scripts vs no scripts correctly.

apoelstra replied:
No, we should not enforce it at the type level. If users fail to do this then I hope this library would still work for them.

sanket1729 replied:
This is being done in #691 . I will update this PR with Strong types after that is merged.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,777 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// constructing the tree as a Huffman tree is the optimal way to minimize average
+    /// case satisfaction cost. This function takes input a list of tuple(usize, &Script)
+    /// where usize represents the satisfaction weights of the branch.
+    /// For example, [(3, S1), (2, S2), (5, S3)] would construct a TapTree that has optimal
```
This is a vector of tuples, not a BtreeMap/HashMap

dr-orlovsky replied:
Yes, but frequently iterator over tuples is constructed from map types. Should't we prevent/discourage this here?

apoelstra replied:
I think this method is fine. It clearly documents its expectations. If the user provides an iterator which does not iterate in the correct order, we cannot help them.

sanket1729 commented:
@dr-orlovsky No issues, your comments are helpful. It did help in catching an interesting caching bug that I and a few other reviewers missed! I have addressed all of your comments. Awaiting your review

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+    /// Create a new BIP341 [`TapTweakHash`] from key and tweak
+    /// Produces H_taptweak(P||R) where P is internal key and R is the merkle root
+    pub fn from_key_and_tweak(
```
nit: Rust API guidelines recomend to name functions with multiple params not `from_*` but `with*` (I know not everyone agree with that) - or simply `new` (since we do not have other constructors here)

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+
+impl TapLeafHash {
+    /// function to compute leaf hash from components
+    pub fn from_script(script: &Script, ver: LeafVersion) -> TapLeafHash {
```
Ibid

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        let mut eng = TapTweakHash::engine();
+        // always hash the key
+        eng.input(&internal_key.serialize());
+        if let Some(h) = merkle_root {
+            eng.input(&h);
+        } else {
+            // nothing to hash
+        }
```
nit: 
```suggestion
        `merkle_root.map(|h| eng.input(h))`
```

sanket1729 replied:
It might be a matter of personal taste, but I don't like using `map` when it mutates the external environmental variables (eng in this case). I see map as converting Option<T> -> Option<U>. 

In cases where there is an environmental mutation, I prefer using loops/explicit if let over functional methods like `iter(), map()`. In other words, I only like functional methods when there are no side effects.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        let mut eng = TapLeafHash::engine();
+        ver.as_u8()
+            .consensus_encode(&mut eng)
+            .expect("engines don't err");
```
nit:
```suggestion
            .expect("TapLeafHash engine broken on tapleaf version serialization");
```

apoelstra replied:
Honestly I'd rather this be an `.unwrap()` with a comment. All these unique strings will eventually add up to nontrivial library bloat, for code which will literally never be exercised.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            .expect("engines don't err");
+        script
+            .consensus_encode(&mut eng)
+            .expect("engines don't err");
```
nit:
```suggestion
            .expect("TapLeafHash engine broken on script serialization");
```

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// The sign final output pubkey as per BIP 341
+    output_key_parity: bool,
+    /// The tweaked output key
+    output_key: schnorr::PublicKey,
```
Should we wait for #691 to be merged first such that we can use a proper key type here?

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    // Internal function to compute [`TaprootSpendInfo`] from NodeInfo
+    fn from_node_info<C: secp256k1::Verification>(
```
nit: The same as above on using `with_*` instead of `from_*` if the function takes more than one param

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TaprootBuilder {
+    // The following doc-comment is from bitcoin core, but modified for rust
+    // The current state of the builder.
```
Can't parse the meaning of this line of the comment and how it is related to the previous line (the same sentence or a new one)
```suggestion
```

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    // Create a new leaf with NodeInfo
+    fn new_leaf_with_ver(script: Script, ver: LeafVersion) -> Self {
```
nit: `with_leaf_ver` more ideomatic

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    // Combine two NodeInfo's to create a new parent
+    fn combine(a: Self, b: Self) -> Result<Self, TaprootBuilderError> {
```
Why the function does not take `&mut self` as first argument? This will make calling it above in line https://github.com/rust-bitcoin/rust-bitcoin/pull/677/files#diff-aa9330b8acaba59f1d0418de0acdd5836506b48566d28886e2a3558e1e19de1bR467 simpler: `node.combine(child)?;`

sanket1729 replied:
I guess this is about how we view the tree structure. I think conceptually it makes more sense to consume both the children and create a parent. I believe this helps in understanding the function source code easily whereas the other might be slightly more complicated to reason about.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+// Internally used structure to store information about taproot leaf node
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+struct LeafInfo {
+    /// The underlying script
```
nit: we should either use doc-style or normal comments at both struct and fields

tcharding replied:
Normal comments because its a private struct, right?

Kixunil replied:
I believe the point here is consistency and I prefer using doc comments. They will be only visible if doc is generated with special flags.

tcharding replied:
Oh wow, I have been doing patch review for the last two years telling folks not to use `///` for private structs based on the docs build systems in other languages. Thanks for schooling me.

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+/// The Merkle proof for inclusion of a tree in a taptree hash
+// The type of hash is sha256::Hash because the vector might contain
+// both TapBranchHash and TapLeafHash
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
```
`Default`?

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    /// Serialize to a writer. Returns the number of bytes written
+    pub fn serialize_to_writer<Write: io::Write>(&self, mut writer: Write) -> io::Result<usize> {
```
nit: maybe simply `pub fn write`, or `encode`?

dr-orlovsky commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    }
+
+    /// Serialize to a writer. Returns the number of bytes written
+    pub fn serialize_to_writer<Write: io::Write>(&self, mut writer: Write) -> io::Result<usize> {
```
nit on naming, ibid as above

dr-orlovsky reviewed (approved):
LGTM now and utACK 89d96c0b1b8e6233af00eb551a737ae1d7c26747 mod several nits, which should not block mergning.

However lets consider whether we need to wait taptweaked key type from #691 before the merge or we need to re-base #691 after this is merged.

GeneFerneau commented:
tACK [89d96c0](https://github.com/rust-bitcoin/rust-bitcoin/pull/677/commits/89d96c0b1b8e6233af00eb551a737ae1d7c26747)

sanket1729 commented:
@dr-orlovsky addressed all of your comments. I think if we get another review for this. We should just merge this and proceed with other taproot stuff. I can work on fixing up and making sure changes in #691 are reflected here. 

Finally, I think about the API guidelines, I think we should have a broader discussion. My rule is to always follow the current convention in the codebase. Using  `new_*` and `from_*` over `with_*`. If we both(with and new) of these it would be even worse. If we want to switch to with_ (assuming we have consensus), we should do it for all APIs in a non-breaking way in a future release.

Kixunil commented:
My understanding of naming:

* `new` - obvious constructor, may be empty (`String::new()`, `Box::new(T)`)
* `from_` - takes *same* information as `new` just in different representation (e.g. `String::from_utf8(Vec<u8>)`)
* `with_` - takes *additional* information along with what `new` takes `String::with_capacity(usize)`

tcharding commented:
> My understanding of naming:
> 
>     * `new` - obvious constructor, may be empty (`String::new()`, `Box::new(T)`)
>     * `from_` - takes _same_ information as `new` just in different representation (e.g. `String::from_utf8(Vec<u8>)`)
>     * `with_` - takes _additional_ information along with what `new` takes `String::with_capacity(usize)`

I thought `String` was a bit of a special case but a quick look at the std lib shows its not. Perhaps conventions have changed for more recent code? I was under the impression that
- Constructors with zero args should use `Default`
- Constructors with one arg should use `From<Foo>` (unless the foo needs further restriction like the `String;:from_utf8` case)
- Constructors with multiple args should use `new`.
- Constructors using the builder pattern should use `with(self, foo: Foo)`

I've never managed to work out the pattern for times when we need multiple constructors with multiple args? 

Amusingly I was looking at all the instances of single arg constructors for `Script` yesterday wondering why they used the pattern `new_`.

I don't mean to bike shed here, I'm genuinely interested in these sorts of conventions.

JeremyRubin reviewed (commented):
did some light review

sanket1729 commented on src/util/sighash.rs:
```diff
@@ -88,15 +90,14 @@ pub enum Prevouts<'u> {
     script: &'s Script,
     code_separator_pos: u32,
-    leaf_version: u8,
+    leaf_version: LeafVersion,
```
This is not a part of the current PR under discussion. But I like the name TapScriptPath over ScriptPath

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    TAPROOT_CONTROL_BASE_SIZE + TAPROOT_CONTROL_NODE_SIZE * TAPROOT_CONTROL_MAX_NODE_COUNT;
+
+// type alias for versioned tap script corresponding merkle proof
+type ScriptMerkleProofMap = BTreeMap<(Script, LeafVersion), BTreeSet<TaprootMerkleBranch>>;
```
Yes, this is to allow > 1 instances of a given script. In practice, this should never be used, but since descriptors spec does not disallow it, we have to support it.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    ///
+    /// # Errors:
+    ///
+    /// - When the optimal huffman tree has a depth more than 128
```
This is an improvement. I can create a good first issue for this. If someone really wants this, they are welcome to implement this and I am happy to review it.

apoelstra replied:
If the optimal huffman tree has depth exceeding 128 it means the user has set insane weights. We should error rather than silently doing the wrong thing.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+            // Combine the last two elements and insert a new node
+            let (p1, s1) = node_weights.pop().expect("len must be at least two");
+            let (p2, s2) = node_weights.pop().expect("len must be at least two");
+            // Insert the sum of first two in the tree as a new node
```
Good catch!

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    {
+        let mut node_weights = BinaryHeap::<(usize, NodeInfo)>::new();
+        for (p, leaf) in script_weights {
+            node_weights.push((p, NodeInfo::new_leaf_with_ver(leaf, LeafVersion::default())));
```
I don't follow. Why should we fail if p is repeated? Do you mean if the script is repeated? 

I don't see a reason to fail, having repeated scripts is perfectly fine. Just seems like an additional effort to check for repeat scripts without too much benefit for it.

JeremyRubin replied:
right -- i missed that it's a BinaryHeap and not a BTree, so you can have >1 instance of same element.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    /// each branch being taken. If probabilities/weights for each condition are known,
+    /// constructing the tree as a Huffman tree is the optimal way to minimize average
+    /// case satisfaction cost. This function takes input a list of tuple(usize, &Script)
+    /// where usize represents the satisfaction weights of the branch.
```
I don't follow. There is no multiplication involved here.
```
max_satisfaction_cost = 1(version/output key parity) + 32(internal_key) + 32*(merkle_path_len) + script_satisfaction_cost.
```

Given our input, `script_satisaction_cost` is fixed and we want to optimize merkle_path_len

JeremyRubin replied:
i'm a bit confused tbh. maybe i'm missing some bayes thing.

mostly i'm asking if we should be inferring weights from the cost of the script_satisfaction_cost (which includes script length?)?

An example of where this might come up is if you suppose node A and node B have similar relative odds but B's computes to slightly worse than A's, however for the script underneath B is simple (e.g. a single key) and A's script is quite large (e.g., 100 hash reveals and a key).

Even though B is relative odds marginally less likely, the weighting tells us that B should be cheaper and we'd possibly prefer B to A? It feels like this gets complicated by the issue that a node B could represent a subtree of e.g. a bunch of different satisfactions of different costs but the issue generalizes.

sanket1729 commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+    pub fn new_key_spend<C: secp256k1::Verification>(
+        secp: &Secp256k1<C>,
+        internal_key: schnorr::PublicKey,
+        merkle_root: Option<TapBranchHash>,
```
The return type is `TaprootSpendInfo`

JeremyRubin replied:
ah i misread as TapBranchHash

apoelstra commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        }
+        // Every iteration of the loop reduces the node_weights.len() by exactly 1
+        // Therefore, the loop will eventually terminate with exactly 1 element
+        debug_assert!(node_weights.len() == 1);
```
Is there *any* circumstance where nontermination is preferable to a visible crash?

JeremyRubin replied:
generally, non termination is considered safe in rust https://doc.rust-lang.org/reference/behavior-not-considered-unsafe.html, but undesirable. The existing version here -- by virtue of using a while loop -- might not terminate as well if there's a logic bug in the loop. There is a guaranteed to terminate version if you want to ensure the logic ends:

```
for i in 0..(2*node_weights.len()) {
if node_weights.len() == 1 {
  break;
}
}
```

but this doesn't eliminate the panic necessarily.

Kixunil commented on src/util/taproot.rs:
```diff
@@ -82,12 +93,804 @@ sha256t_hash_newtype!(TapSighashHash, TapSighashTag, MIDSTATE_TAPSIGHASH, 64,
+        internal_key: schnorr::PublicKey,
+        script_weights: I,
+    ) -> Result<Self, TaprootBuilderError>
+    where
```
Probability is usually expressed as float in range `0..=1`, so it may be unintuitive/confusing to use integer. Perhaps worth a newtype with some helper methods? Or maybe more details explaining how would one compute the numbers?

apoelstra commented:
This PR has exceeded Github's ability to reasonably display comments. I would prefer future reviews to be in the form of followup PRs.

Suggestions for non-github platforms are also welcome but that's a long term thing.

sanket1729 commented:
So far, I think the only things I see remaining are:

- A doc-comment regarding explaining the usage of u64 for odds over probability. 
- The general API guidelines about the `new_`, `with_`. We basically decided to do this later and is being discussed on #692. 

Anything else that I am missing? If there is nothing critical I would request to proceed on this and I can address all of the nits in a follow up PR.

apoelstra reviewed (approved):
ACK fa8c3f6e444bf2699e3be707a7890b9e22fa757b

sanket1729 commented:
@GeneFerneau @RCasatta, @dr-orlovsky, sorry for dismissing your Acks. if you are interested in taking another look.

dr-orlovsky reviewed (approved):
ACK fa8c3f6e444bf2699e3be707a7890b9e22fa757b
